File: stripe-webhook-not-delivered-configuration/node/check-webhook.js
/**
 * Check whether the Stripe webhook that WooCommerce depends on is set up right.
 * Read only. It reports problems, it does not change anything.
 *
 * Guide: https://www.allanninal.dev/woocommerce/stripe-webhook-not-delivered-configuration/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const STORE_HOST = new URL(process.env.WOO_STORE_URL || "https://example.com").host;

const REQUIRED = ["payment_intent.succeeded", "payment_intent.payment_failed",
                  "charge.succeeded", "charge.refunded"];

export function endpointHealth(endpoint, storeHost) {
  const events = new Set(endpoint.enabled_events || []);
  const covers = events.has("*") || REQUIRED.every((e) => events.has(e));
  let hostOk = false;
  try { hostOk = new URL(endpoint.url).host === storeHost; } catch (e) { hostOk = false; }
  return {
    id: endpoint.id,
    url: endpoint.url || "",
    enabled: endpoint.status === "enabled",
    pointsAtStore: hostOk,
    coversEvents: covers,
    missingEvents: covers ? [] : REQUIRED.filter((e) => !events.has(e)),
  };
}

async function checkEndpoints(storeHost) {
  const reports = [];
  for await (const endpoint of stripe.webhookEndpoints.list({ limit: 100 })) {
    reports.push(endpointHealth(endpoint, storeHost));
  }
  return reports;
}

async function undeliveredRecent(limit = 100) {
  let pending = 0, total = 0;
  for await (const event of stripe.events.list({ limit })) {
    total++;
    if ((event.pending_webhooks || 0) > 0) pending++;
  }
  return { checked: total, stillPending: pending };
}

export async function run() {
  const reports = await checkEndpoints(STORE_HOST);
  const healthy = reports.filter((r) => r.enabled && r.pointsAtStore && r.coversEvents);
  console.log(`Found ${reports.length} endpoint(s). ${healthy.length} healthy for this store.`);
  for (const r of reports) {
    const flags = [];
    if (!r.enabled) flags.push("disabled");
    if (!r.pointsAtStore) flags.push("wrong domain");
    if (!r.coversEvents) flags.push("missing events: " + r.missingEvents.join(", "));
    console.log(`  ${r.id}  ${r.url}  ->  ${flags.length ? "PROBLEM: " + flags.join("; ") : "OK"}`);
  }
  if (!healthy.length) console.log("No healthy endpoint points at this store. That is why orders do not update.");
  const delivery = await undeliveredRecent();
  console.log(`Recent events checked: ${delivery.checked}, still pending delivery: ${delivery.stillPending}`);
  if (delivery.stillPending) {
    console.log("Events are not being accepted by your endpoint. Check for a firewall, CDN, or a server error.");
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: stripe-webhook-not-delivered-configuration/python/check_webhook.py
"""Check whether the Stripe webhook that WooCommerce depends on is set up right.
Read only. It reports problems, it does not change anything.

Guide: https://www.allanninal.dev/woocommerce/stripe-webhook-not-delivered-configuration/
"""
import os
from urllib.parse import urlparse
import stripe

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
STORE_HOST = urlparse(os.environ["WOO_STORE_URL"]).netloc

REQUIRED = {"payment_intent.succeeded", "payment_intent.payment_failed",
            "charge.succeeded", "charge.refunded"}


def endpoint_health(endpoint, store_host):
    events = set(endpoint.get("enabled_events", []))
    covers = "*" in events or REQUIRED.issubset(events)
    host_ok = urlparse(endpoint.get("url", "")).netloc == store_host
    return {
        "id": endpoint["id"],
        "url": endpoint.get("url", ""),
        "enabled": endpoint["status"] == "enabled",
        "points_at_store": host_ok,
        "covers_events": covers,
        "missing_events": sorted(REQUIRED - events) if not covers else [],
    }


def check_endpoints(store_host):
    return [endpoint_health(e, store_host)
            for e in stripe.WebhookEndpoint.list(limit=100).auto_paging_iter()]


def undelivered_recent(limit=100):
    pending = total = 0
    for event in stripe.Event.list(limit=limit).auto_paging_iter():
        total += 1
        if event.get("pending_webhooks", 0) > 0:
            pending += 1
    return {"checked": total, "still_pending": pending}


def run():
    reports = check_endpoints(STORE_HOST)
    healthy = [r for r in reports if r["enabled"] and r["points_at_store"] and r["covers_events"]]
    print(f"Found {len(reports)} endpoint(s). {len(healthy)} healthy for this store.")
    for r in reports:
        flags = []
        if not r["enabled"]:
            flags.append("disabled")
        if not r["points_at_store"]:
            flags.append("wrong domain")
        if not r["covers_events"]:
            flags.append("missing events: " + ", ".join(r["missing_events"]))
        status = "OK" if not flags else "PROBLEM: " + "; ".join(flags)
        print(f"  {r['id']}  {r['url']}  ->  {status}")
    if not healthy:
        print("No healthy endpoint points at this store. That is why orders do not update.")
    delivery = undelivered_recent()
    print(f"Recent events checked: {delivery['checked']}, still pending delivery: {delivery['still_pending']}")
    if delivery["still_pending"]:
        print("Events are not being accepted by your endpoint. Check for a firewall, CDN, or a server error.")


if __name__ == "__main__":
    run()
